File: Station.java
package com.itouchchina.metro;

import java.util.Map;
import java.util.Set;
import java.util.TreeMap;
import java.util.TreeSet;

public class Station {
	
	private static final int transTime = 100;//换乘时间，单位是秒，必须>0，不同线路之间换乘暂时认为是统一的换乘时间
	
	private String name;//站名
	
	private String id;
	
	/**
	 * 节点Map<到站时间, 节点>
	 */
	private Map<Long, Node> nodeMap = new TreeMap<Long, Node>();
	
	/**
	 * 所有节点
	 */
	private Node[] nodes = null;

	public Station(String id, String name){
		this.id = id;
		this.name = name;
	}
	
	public String getName(){
		return name;
	}
	
	public String getId(){
		return id;
	}
	
	public Node setNode(int arriveTime, int routeId, int tripId){
		long key = arriveTime * 50000 + tripId;
		Node node = nodeMap.get(key);
		if(node == null){
			node = new Node(this, arriveTime, routeId, tripId);
			nodeMap.put(key, node);
		}
		return node;
	}
	
	public int getNodeCount(){
		return nodes.length;
	}
	
	public Node getNode(int index){
		return nodes[index];
	}
	
	public int getTransTime(int route_from, int route_to){
		if(route_from == route_to){
			return 0;
		}
		return Station.transTime;
	}
	
	/**
	 * 对站内等待或者换乘的节点进行连接
	 * 如果两个节点是同一站，同一线路，不需要考虑换乘时间
	 * 如果两个节点是同一站的不同路线，需要考虑换乘时间
	 * 对于同一路线只连接最相邻的两个节点
	 * 对于不同路线只连接最相邻的两个节点
	 */
	public void selfLink(){
		nodes = nodeMap.values().toArray(new Node[0]);
		Set<Integer> routes = new TreeSet<Integer>();
		for(int i = 0;i<nodes.length;i++){
			routes.add(nodes[i].getRouteId());
		}
		for (int i = 0; i < nodes.length; i++) {
			Node from = nodes[i];
			for (int route : routes) {
				int trans = getTransTime(from.getRouteId(), route);
				for (int j = 0; j < nodes.length; j++) {
					Node to = nodes[j];
					if(to.getRouteId() == route){
						if(to.getTime() > from.getTime() + trans){
							Node.link(from, to);
							break;
						}
					}
				}
			}
		}
		
	}
	
	public void init(){
		for(int i = 0;i<nodes.length;i++){
			nodes[i].init();
		}
	}
		
	public Node getBestNode(){
		for(int i = 0;i<nodes.length;i++){
			if(nodes[i].isVisited()){
				return nodes[i];
			}
		}
		return null;
	}
}

File: Station.m
//
//  Station.m
//  tokyo-metro
//
//  Created by llch on 15/10/12.
//  Copyright © 2015年 llch. All rights reserved.
//

#import "Station.h"

@interface Station ()
{
    int _transTime;
}

@end

@implementation Station

-(instancetype) initWithSting:(NSString *)station_id Name:(NSString *)name{
    self = [super init];
    if (self){
        _name = [NSString stringWithString:name];
        _stationId = [NSString stringWithString:station_id];
        _transTime = 100;
        _nodeMap = [[NSMutableDictionary alloc] initWithCapacity:0];
        _nodes = [[NSMutableArray alloc]  initWithCapacity:0];
    }
    return  self;
}
-(Node *)setNodeArriveTime:(int)arriveTime routId:(int)routeId tripId:(int)tripId{
    long key = arriveTime * 50000 + tripId;
    Node * node  = [_nodeMap objectForKey:[NSNumber numberWithLong:key]];
    if (node == nil) {
        node = [[Node alloc] initWithStationg:self Time:arriveTime routeId:routeId tripId:tripId];
        [_nodeMap setObject:node forKey:[NSNumber numberWithLong:key]];
    }
    return node;
}
-(int)getNodeCount{
    return (int)_nodes.count;
}
-(Node *)getNode:(int) index{
    return [_nodes objectAtIndex:index];
}
-(int)getTransTimeFrom:(int) route_from To:(int) rout_to{
    if (route_from == rout_to) {
        return 0;
    }
    return _transTime;
}
-(void)selfLink{
    NSArray * keys = [_nodeMap allKeys];
    NSArray * sortedKeys = [keys sortedArrayUsingComparator:^NSComparisonResult(id  _Nonnull obj1, id  _Nonnull obj2) {
        long key1 = [obj1 longValue];
        long key2 = [obj1 longValue];
        if (key1 > key2) {
            return NSOrderedDescending;
        }
        if (key1 < key2) {
            return NSOrderedAscending;
        }
        return NSOrderedSame;
    }];
    [_nodes removeAllObjects];
    for (id key in sortedKeys) {
        [_nodes addObject:[_nodeMap objectForKey:key]];
    }
    NSMutableOrderedSet * routes = [[NSMutableOrderedSet alloc] initWithCapacity:0];
    for (int i = 0; i < _nodes.count; i ++) {
        [routes addObject:[NSNumber numberWithInt:((Node *)_nodes[i]).routeId]];
    }
    [routes sortUsingComparator:^NSComparisonResult(id  _Nonnull obj1, id  _Nonnull obj2) {
        int n1 = [obj1 intValue];
        int n2 = [obj2 intValue];
        if (n1 > n2) {
            return NSOrderedDescending;
        }
        if (n1 < n2) {
            return NSOrderedAscending;
        }
        return NSOrderedSame;
    }];
    if ([_name isEqualToString:@"大手町"]) {
        NSLog(@"大手町routs count:%lu",(unsigned long)[routes count]);
    }
    for (int i = 0 ; i < _nodes.count; i ++) {
        Node * from = _nodes[i];
        for (NSNumber * routNumber in routes) {
            int rout = [routNumber intValue];
            int trans = [self getTransTimeFrom:from.routeId To:rout];
            for (int j = 0; j < _nodes.count; j++) {
                Node * to = _nodes[j];
                if (to.routeId == rout) {
                    if (to.time > (from.time + trans)) {
                        [Node linkfrom:from To:to];
                        break;
                    }
                }
            }
        }
    }
}
-(void)setUpinit{
    for (int i = 0; i < _nodes.count; i++) {
        [((Node *)_nodes[i]) setUpinit];
    }
}
-(Node *)getBestNode{
    for (int i = 0; i< _nodes.count; i++) {
        Node * node = (Node *)_nodes[i];
        if (node.visited) {
            return node;
        }
    }
    return nil;
}
@end
